core/src/toga/images.py

```diff
@@ -1,52 +1,104 @@
 from __future__ import annotations
 
+from typing import Any
+
 from pathlib import Path
 
 import toga
 from toga.platform import get_platform_factory
 
+import time
+import os
 
 class Image:
     def __init__(
         self,
         path: str | None | Path = None,
         *,
         data: bytes | None = None,
+        pil_image: Any | None = None,
     ):
         """Create a new image.
 
-        An image must be provided either a ``path`` or ``data``, but not both.
+        An image must be one of ``path``, ``data`` or ``pil_image``
 
         :param path: Path to the image to load. This can be specified as a string, or as
             a :any:`pathlib.Path` object. The path can be an absolute file system path,
             or a path relative to the module that defines your Toga application class.
         :param data: A bytes object with the contents of an image in a supported format.
+        :param pil_image: PIL.Image object created from an image of a supported format.
         :raises FileNotFoundError: If a path is provided, but that path does not exist.
         :raises ValueError: If the path or data cannot be loaded as an image.
+        :raises TypeError: If pil_image is provided but the type of the object is not PIL.Image
         """
-        if path is None and data is None:
-            raise ValueError("Either path or data must be set.")
-        if path is not None and data is not None:
-            raise ValueError("Only either path or data can be set.")
-
+        # At first we will create a list with these three variable, then count how many None is in that list.
+        # If the number of None is 1 -> raises ValueError. One and only One of the three variables must be set but here two of them are set
+        # If the number of None is 3 -> raises ValueError. One and only One of the three variables must be set but here none of them are set 
+        # If the number of None is 2 -> Ok. Check the validity of the value of that non-None variable
+        none_count = [path, data, pil_image].count(None)
+        if none_count != 2:
+            raise ValueError("One and Only one of the three args (path, data, pil_image) must be set.")
+        # checking validity of the arg(one of the three)
         if path is not None:
             if isinstance(path, Path):
                 self.path = path
             else:
                 self.path = Path(path)
             self.data = None
-        else:
-            self.path = None
+            self.pil_image = None
+        if data is not None:
             self.data = data
+            self.path = None
+            self.pil_image = None
+        if pil_image is not None:
+            try:
```

Comment:
You don't want to be doing the import every time; we can do the import at the top of the page, and if there's an ImportError, store `PIL_Image` as None. Then, in the later code, you can check for whether PIL_Image is None and raise an exception.